spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/FilteringBatchMessageListenerAdapter.java

```diff
@@ -79,19 +83,21 @@ public void onMessage(List<ConsumerRecord<K, V>> records, @Nullable Acknowledgme
 
 		List<ConsumerRecord<K, V>> consumerRecords = getRecordFilterStrategy().filterBatch(records);
 		Assert.state(consumerRecords != null, "filter returned null from filterBatch");
-		boolean consumerAware = this.delegateType.equals(ListenerType.ACKNOWLEDGING_CONSUMER_AWARE)
-						|| this.delegateType.equals(ListenerType.CONSUMER_AWARE);
-		/*
-		 *  An empty list goes to the listener if ackDiscarded is false and the listener can ack
-		 *  either through the acknowledgment
-		 */
-		if (consumerRecords.size() > 0 || consumerAware
-				|| (!this.ackDiscarded && this.delegateType.equals(ListenerType.ACKNOWLEDGING))) {
-			invokeDelegate(consumerRecords, acknowledgment, consumer);
+
+		if (consumerRecords.isEmpty()) {
+			if (acknowledgment != null) {
+				invokeDelegate(consumerRecords, acknowledgment, consumer);
+			}
 		}
 		else {
-			if (this.ackDiscarded && acknowledgment != null) {
-				acknowledgment.acknowledge();
+			if (this.consumerAware
+				|| (!this.ackDiscarded && this.delegateType.equals(ListenerType.ACKNOWLEDGING))) {
+				invokeDelegate(consumerRecords, acknowledgment, consumer);
+			}
+			else {
+				if (this.ackDiscarded && acknowledgment != null) {
+					acknowledgment.acknowledge();
```

Comment:
I don't think this is correct.
We are in this `else` branch when `!consumerRecords.isEmpty()`, therefore in my understanding we always have to call `invokeDelegate()`.
It sounds more like we are looking for something like this:
```
	if (!consumerRecords.isEmpty() || acknowledgment != null || consumerAware
				|| (!this.ackDiscarded && this.delegateType.equals(ListenerType.ACKNOWLEDGING))) {
			invokeDelegate(consumerRecords, acknowledgment, consumer);
```

However I believe that `this.delegateType.equals(ListenerType.ACKNOWLEDGING)` covers `acknowledgment != null`.

I probably need to look close what is the request in that issue.

Reply:
> However I believe that this.delegateType.equals(ListenerType.ACKNOWLEDGING) covers acknowledgment != null.

IMHO, Thus, `acknowledgment != null` has more narrow than `this.delegateType.equals(ListenerType.ACKNOWLEDGING)`. 
`delegateType` is determined when spring context started with this codes.
https://github.com/spring-projects/spring-kafka/blob/5962e756814df74222368c73dbde6fbd578f9e9e/spring-kafka/src/main/java/org/springframework/kafka/listener/ListenerUtils.java#L54-L76

However, `Acknowledgment` instance is created only `isAnyManualAck == true`. 

https://github.com/spring-projects/spring-kafka/blob/5962e756814df74222368c73dbde6fbd578f9e9e/spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java#L2380-L2403

`isAnyManualAck` is determined on `Constructor`.
`isAnyManualAck` consider only `AckMode.MANUAL` and `AckMode.MANUAL_IMMEDIATE`.

https://github.com/spring-projects/spring-kafka/blob/5962e756814df74222368c73dbde6fbd578f9e9e/spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java#L853-L855

Thus, `acknowledgment != null` has more narrow than `this.delegateType.equals(ListenerType.ACKNOWLEDGING)`.

Reply:
Why we should have to call `invokeDelegate()` always? 
It loos like that `KafkaMessageListener` will ack batch eventually when(`AckMode != MANUAL_IMMEDIATE`) and (`AckMODE != MANUAL`) and (`!autoCommit`) condition is satisfied,

https://github.com/spring-projects/spring-kafka/blob/5962e756814df74222368c73dbde6fbd578f9e9e/spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java#L2301-L2322

IMHO, From codes above, If `consumerRecords.isEmpty()` is `true`, we don't need to invoke listener always. 
However, Condition which you said (it looks like that same with current code on main branch, `!consumerRecords.isEmpty()` equals `consumerRecords.size() > 0 `, right?), yes, we have to call `invoke` always.
```java
if (!consumerRecords.isEmpty() || acknowledgment != null || consumerAware || (!this.ackDiscarded && this.delegateType.equals(ListenerType.ACKNOWLEDGING))) { invokeDelegate(consumerRecords, acknowledgment, consumer);
```